File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
    if (!matrix.length || !matrix[0].length) return false;

    let n = matrix[0].length;

    let left = 0, right = matrix.length * n - 1;

    while (left <= right) {
        let mid = Math.floor((left + right) / 2);

        let i = Math.floor(mid / n);
        let j = mid % n;

        if (matrix[i][j] === target) return true;
        else if (matrix[i][j] > target) right = mid - 1;
        else left = mid + 1;
    }

    return false;
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m=matrix.size();
        int n=matrix[0].size();
        
        int length=m*n;
        int left=0;
        int right=length-1;
        while(left<=right)
        {
            int mid=(left+right)/2;
            int i=mid/n;
            int j=mid%n;
            if(matrix[i][j]==target)
                return true;
            else if(matrix[i][j]<target)
            {
                left=mid+1;
            }
            else
                right=mid-1;
        }
        return 0;
    }
};